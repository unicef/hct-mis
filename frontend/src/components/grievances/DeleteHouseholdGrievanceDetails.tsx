import { Box, Button, Grid, Typography } from '@material-ui/core';
import { Field, Formik } from 'formik';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { useBusinessArea } from '../../hooks/useBusinessArea';
import { useSnackbar } from '../../hooks/useSnackBar';
import { FormikRadioGroup } from '../../shared/Formik/FormikRadioGroup';
import { FormikTextField } from '../../shared/Formik/FormikTextField';
import { GRIEVANCE_TICKET_STATES } from '../../utils/constants';
import { choicesToDict } from '../../utils/utils';
import {
  GrievanceTicketDocument,
  GrievanceTicketQuery,
  useApproveDeleteHouseholdDataChangeMutation,
  useHouseholdChoiceDataQuery,
} from '../../__generated__/graphql';
import { useConfirmation } from '../core/ConfirmationDialog';
import { ContentLink } from '../core/ContentLink';
import { LabelizedField } from '../core/LabelizedField';
import { LoadingComponent } from '../core/LoadingComponent';
import { Title } from '../core/Title';
import { ApproveBox } from './GrievancesApproveSection/ApproveSectionStyles';

export const DeleteHouseholdGrievanceDetails = ({
  ticket,
  canApproveDataChange,
}: {
  ticket: GrievanceTicketQuery['grievanceTicket'];
  canApproveDataChange: boolean;
}): React.ReactElement => {
  const { t } = useTranslation();
  const { showMessage } = useSnackbar();
  const businessArea = useBusinessArea();
  const [mutate] = useApproveDeleteHouseholdDataChangeMutation();
  const confirm = useConfirmation();
  const {
    data: choicesData,
    loading: choicesLoading,
  } = useHouseholdChoiceDataQuery();

  const isForApproval = ticket.status === GRIEVANCE_TICKET_STATES.FOR_APPROVAL;
  const approveEnabled = isForApproval;
  const dialogText = t(
    'You did not approve the following household to be withdrawn. Are you sure you want to continue?',
  );

  const dialogContentWithdrawReason = (
    <Formik
      initialValues={{ withdrawReason: '', duplicateHouseholdId: null }}
      enableReinitialize
      onSubmit={(values) => {
        console.log('VALUES', values);
      }}
    >
      {({ submitForm }) => (
        <Box display='flex' flexDirection='column'>
          <Box mt={2}>
            <Typography variant='body2'>
              {t('Please provide the reason of withdrawal of this household.')}
            </Typography>
          </Box>
          <Field
            name='withdrawReason'
            choices={[
              {
                value: 'duplicate',
                name: 'This household is a duplicate of another household',
              },
            ]}
            component={FormikRadioGroup}
            noMargin
          />
          <Grid container>
            <Grid item xs={6}>
              <Field
                name='duplicateHouseholdId'
                fullWidth
                variant='outlined'
                label={t('Household Id')}
                component={FormikTextField}
                required
              />
            </Grid>
          </Grid>
          <Field
            name='withdrawReason'
            choices={[{ value: 'other', name: 'Other' }]}
            component={FormikRadioGroup}
            noMargin
          />
          <Button onClick={submitForm}>Submit</Button>
        </Box>
      )}
    </Formik>
  );

  const { approveStatus } = ticket.deleteHouseholdTicketDetails;

  if (choicesLoading) return <LoadingComponent />;

  const residenceChoicesDict = choicesToDict(
    choicesData.residenceStatusChoices,
  );

  return (
    <ApproveBox>
      <Title>
        <Box display='flex' justifyContent='space-between'>
          <Typography variant='h6'>{t('Household to be withdrawn')}</Typography>
          {canApproveDataChange && (
            <Button
              onClick={() =>
                confirm({
                  title: approveStatus
                    ? t('Warning')
                    : t('Reason of withdrawal'),
                  content: approveStatus
                    ? dialogText
                    : dialogContentWithdrawReason,
                  continueText: t('Confirm'),
                }).then(async () => {
                  try {
                    await mutate({
                      variables: {
                        grievanceTicketId: ticket.id,
                        approveStatus: !ticket.deleteHouseholdTicketDetails
                          ?.approveStatus,
                      },
                      refetchQueries: () => [
                        {
                          query: GrievanceTicketDocument,
                          variables: { id: ticket.id },
                        },
                      ],
                    });
                    if (approveStatus) {
                      showMessage(t('Changes Disapproved'));
                    }
                    if (!approveStatus) {
                      showMessage(t('Changes Approved'));
                    }
                  } catch (e) {
                    e.graphQLErrors.map((x) => showMessage(x.message));
                  }
                })
              }
              variant={approveStatus ? 'outlined' : 'contained'}
              color='primary'
              disabled={!approveEnabled}
            >
              {approveStatus ? t('Disapprove') : t('Approve')}
            </Button>
          )}
        </Box>
      </Title>
      <Grid container spacing={6}>
        <Grid item xs={3}>
          <LabelizedField label={t('Household Size')}>
            {ticket.household.size}
          </LabelizedField>
        </Grid>
        <Grid item xs={3}>
          <LabelizedField label={t('Residence Status')}>
            {residenceChoicesDict[ticket.household.residenceStatus]}
          </LabelizedField>
        </Grid>
        <Grid item xs={3}>
          <LabelizedField label={t('Head of Household')}>
            <ContentLink
              href={`/${businessArea}/population/individuals/${ticket.household.headOfHousehold.id}`}
            >
              {ticket.household.headOfHousehold.fullName}
            </ContentLink>
          </LabelizedField>
        </Grid>
        <Grid item xs={3}>
          <LabelizedField label={t('Country')}>
            {ticket.household.country}
          </LabelizedField>
        </Grid>

        <Grid item xs={3}>
          <LabelizedField label={t('Country of Origin')}>
            {ticket.household.countryOrigin}
          </LabelizedField>
        </Grid>
        <Grid item xs={3}>
          <LabelizedField label={t('Address')}>
            {ticket.household.address}
          </LabelizedField>
        </Grid>
        <Grid item xs={3}>
          <LabelizedField label={t('Village')}>
            {ticket.household.village}
          </LabelizedField>
        </Grid>
        <Grid item xs={3}>
          <LabelizedField label={t('Administrative Level 1')}>
            {ticket.household.admin1?.name}
          </LabelizedField>
        </Grid>
        <Grid item xs={3}>
          <LabelizedField label={t('Administrative Level 2')}>
            {ticket.household.admin2?.name}
          </LabelizedField>
        </Grid>
        <Grid item xs={3}>
          <LabelizedField label={t('Geolocation')}>
            {ticket.household.geopoint
              ? `${ticket.household.geopoint.coordinates[0]}, ${ticket.household.geopoint.coordinates[1]}`
              : '-'}
          </LabelizedField>
        </Grid>
        <Grid item xs={3}>
          <LabelizedField label={t('UNHCR CASE ID')}>
            {ticket.household?.unhcrId}
          </LabelizedField>
        </Grid>
      </Grid>
    </ApproveBox>
  );
};
