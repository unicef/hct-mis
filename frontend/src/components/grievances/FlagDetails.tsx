import {
  Box,
  Button,
  makeStyles,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
  Typography,
} from '@material-ui/core';
import moment from 'moment';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { DATE_FORMAT } from '../../config';
import { GRIEVANCE_TICKET_STATES } from '../../utils/constants';
import {
  GrievanceTicketDocument,
  GrievanceTicketNode,
  useApproveSystemFlaggingMutation,
} from '../../__generated__/graphql';
import { useConfirmation } from '../core/ConfirmationDialog';
import { FlagTooltip } from '../core/FlagTooltip';
import { Title } from '../core/Title';
import { UniversalMoment } from '../core/UniversalMoment';
import { ApproveBox } from './GrievancesApproveSection/ApproveSectionStyles';
import { ViewSanctionList } from './ViewSanctionList';

export function FlagDetails({
  ticket,
  canApproveFlag,
}: {
  ticket: GrievanceTicketNode;
  canApproveFlag: boolean;
}): React.ReactElement {
  const { t } = useTranslation();
  const confirm = useConfirmation();
  const useStyles = makeStyles(() => ({
    table: {
      minWidth: 100,
    },
  }));
  const [approve] = useApproveSystemFlaggingMutation({
    refetchQueries: () => [
      {
        query: GrievanceTicketDocument,
        variables: { id: ticket.id },
      },
    ],
  });
  const classes = useStyles();
  const confirmationText = t(
    'Are you sure you want to confirm flag (sanction list match) ?',
  );
  const removalText = t('Are you sure you want to remove the flag ?');
  const details = ticket.systemFlaggingTicketDetails;
  const isFlagConfirmed = details.approveStatus;
  return (
    <ApproveBox>
      <Title>
        <Box display='flex' justifyContent='space-between'>
          <Typography variant='h6'>{t('Flag Details')}</Typography>
          <Box>
            <ViewSanctionList
              referenceNumber={details.sanctionListIndividual.referenceNumber}
            />
            {canApproveFlag && (
              <Button
                disabled={
                  ticket.status !== GRIEVANCE_TICKET_STATES.FOR_APPROVAL
                }
                onClick={() =>
                  confirm({
                    content: isFlagConfirmed ? removalText : confirmationText,
                  }).then(() =>
                    approve({
                      variables: {
                        grievanceTicketId: ticket.id,
                        approveStatus: !details.approveStatus,
                      },
                    }),
                  )
                }
                variant='outlined'
                color='primary'
              >
                {isFlagConfirmed ? t('REMOVE FLAG') : t('CONFIRM FLAG')}
              </Button>
            )}
          </Box>
        </Box>
      </Title>
      <Table className={classes.table}>
        <TableHead>
          <TableRow>
            <TableCell align='left' />
            <TableCell align='left'>{t('Ref. No. on Sanction List')}</TableCell>
            <TableCell align='left'>{t('Full Name')}</TableCell>
            <TableCell align='left'>{t('Date of Birth')}</TableCell>
            <TableCell align='left'>{t('National Ids')}</TableCell>
            <TableCell align='left'>{t('Source')}</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          <TableRow>
            <TableCell align='left'>
              {isFlagConfirmed ? (
                <FlagTooltip
                  message={t('Sanction List Confirmed Match')}
                  confirmed
                />
              ) : (
                ''
              )}
            </TableCell>
            <TableCell align='left'>-</TableCell>
            <TableCell align='left'>
              {details.goldenRecordsIndividual.fullName}
            </TableCell>
            <TableCell align='left'>
              <UniversalMoment>
                {details.goldenRecordsIndividual.birthDate}
              </UniversalMoment>
            </TableCell>
            <TableCell align='left'>
              {details.goldenRecordsIndividual.documents.edges
                .filter((item) => item.node.type.type === 'NATIONAL_ID')
                .map((item) => item.node.documentNumber)
                .join(', ') || '-'}
            </TableCell>
            <TableCell align='left'>{t('Golden Record')}</TableCell>
          </TableRow>
          <TableRow>
            <TableCell align='left' />
            <TableCell align='left'>
              {details.sanctionListIndividual.referenceNumber}
            </TableCell>
            <TableCell align='left'>
              {details.sanctionListIndividual.fullName}
            </TableCell>
            <TableCell align='left'>
              {details.sanctionListIndividual.datesOfBirth.edges
                .map((item) => moment(item.node.date).format(DATE_FORMAT))
                .join(', ') || '-'}
            </TableCell>
            <TableCell align='left'>
              {details.sanctionListIndividual.documents.edges
                .map((item) => item.node.documentNumber)
                .join(', ') || '-'}
            </TableCell>
            <TableCell align='left'>{t('Sanction List')}</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </ApproveBox>
  );
}
