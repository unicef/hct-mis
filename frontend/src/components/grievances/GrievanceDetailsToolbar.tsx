import { Box, Button } from '@material-ui/core';
import EditIcon from '@material-ui/icons/EditRounded';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { Link, useParams, useHistory } from 'react-router-dom';
import styled from 'styled-components';
import { useBusinessArea } from '../../hooks/useBusinessArea';
import { useSnackbar } from '../../hooks/useSnackBar';
import { MiśTheme } from '../../theme';
import {
  GRIEVANCE_CATEGORIES,
  GRIEVANCE_ISSUE_TYPES,
  GRIEVANCE_TICKET_STATES,
} from '../../utils/constants';
import {
  GrievanceTicketDocument,
  GrievanceTicketQuery,
  useGrievanceTicketStatusChangeMutation,
} from '../../__generated__/graphql';
import { BreadCrumbsItem } from '../core/BreadCrumbs';
import { ButtonDialog } from '../core/ButtonDialog';
import { useConfirmation } from '../core/ConfirmationDialog';
import { LoadingButton } from '../core/LoadingButton';
import { PageHeader } from '../core/PageHeader';

const Separator = styled.div`
  width: 1px;
  height: 28px;
  border: 1px solid
    ${({ theme }: { theme: MiśTheme }) => theme.hctPalette.lightGray};
  margin: 0 28px;
`;

const countApprovedAndUnapproved = (
  data,
): { approved: number; notApproved: number } => {
  let approved = 0;
  let notApproved = 0;
  const flattenArray = data.flat(2);
  flattenArray.forEach((item) => {
    if (item.approve_status === true) {
      approved += 1;
    } else {
      notApproved += 1;
    }
  });

  return { approved, notApproved };
};

export const GrievanceDetailsToolbar = ({
  ticket,
  canEdit,
  canSetInProgress,
  canSetOnHold,
  canSendForApproval,
  canSendBack,
  canClose,
  canAssign,
}: {
  ticket: GrievanceTicketQuery['grievanceTicket'];
  canEdit: boolean;
  canSetInProgress: boolean;
  canSetOnHold: boolean;
  canSendForApproval: boolean;
  canSendBack: boolean;
  canClose: boolean;
  canAssign: boolean;
}): React.ReactElement => {
  const { t } = useTranslation();
  const { id } = useParams();
  const { showMessage } = useSnackbar();
  const businessArea = useBusinessArea();
  const confirm = useConfirmation();
  const history = useHistory();
  const breadCrumbsItems: BreadCrumbsItem[] = [
    {
      title: t('Grievance and Feedback'),
      to: `/${businessArea}/grievance-and-feedback/tickets`,
    },
  ];
  const [mutate, { loading }] = useGrievanceTicketStatusChangeMutation();

  const isNew = ticket.status === GRIEVANCE_TICKET_STATES.NEW;
  const isAssigned = ticket.status === GRIEVANCE_TICKET_STATES.ASSIGNED;
  const isInProgress = ticket.status === GRIEVANCE_TICKET_STATES.IN_PROGRESS;
  const isForApproval = ticket.status === GRIEVANCE_TICKET_STATES.FOR_APPROVAL;
  const isOnHold = ticket.status === GRIEVANCE_TICKET_STATES.ON_HOLD;
  const isClosed = ticket.status === GRIEVANCE_TICKET_STATES.CLOSED;
  const isEditable = !isClosed;

  const isFeedbackType =
    ticket.category.toString() === GRIEVANCE_CATEGORIES.POSITIVE_FEEDBACK ||
    ticket.category.toString() === GRIEVANCE_CATEGORIES.NEGATIVE_FEEDBACK ||
    ticket.category.toString() === GRIEVANCE_CATEGORIES.REFERRAL;

  const getClosingConfirmationExtraTextForIndividualAndHouseholdDataChange = (): string => {
    const householdData =
      ticket.householdDataUpdateTicketDetails?.householdData || {};
    const individualData =
      ticket.individualDataUpdateTicketDetails?.individualData || {};
    const allData = {
      ...householdData,
      ...individualData,
      ...householdData?.flex_fields,
      ...individualData?.flex_fields,
    };
    const excludedKeys = [
      'previous_documents',
      'previous_identities',
      'previous_payment_channels',
      'flex_fields',
    ];

    const filteredData = Object.keys(allData)
      .filter((key) => !excludedKeys.includes(key))
      .reduce((obj, key) => {
        return {
          ...obj,
          [key]: allData[key],
        };
      }, {});

    const { approved, notApproved } = countApprovedAndUnapproved(
      Object.values(filteredData),
    );

    if (!notApproved) {
      return '';
    }

    if (!approved) {
      const rejectionMessage = t(
        `You approved 0 changes, remaining proposed changes will be automatically rejected upon ticket closure.`,
      );
      return rejectionMessage;
    }

    const approvalMessage = `You approved ${approved} change${
      approved > 1 ? 's' : ''
    }. Remaining change requests (${notApproved}) will be automatically rejected.`;
    return approvalMessage;
  };

  const getClosingConfirmationExtraTextForOtherTypes = (): string => {
    const hasApproveOption =
      ticket.category?.toString() === GRIEVANCE_CATEGORIES.DATA_CHANGE ||
      ticket.category?.toString() === GRIEVANCE_CATEGORIES.DEDUPLICATION ||
      ticket.category?.toString() === GRIEVANCE_CATEGORIES.SYSTEM_FLAGGING;

    if (!hasApproveOption) {
      return '';
    }

    const notApprovedDeleteIndividualChanges =
      ticket.issueType?.toString() ===
        GRIEVANCE_ISSUE_TYPES.DELETE_INDIVIDUAL &&
      ticket.deleteIndividualTicketDetails?.approveStatus === false;

    const notApprovedAddIndividualChanges =
      ticket.issueType?.toString() === GRIEVANCE_ISSUE_TYPES.ADD_INDIVIDUAL &&
      ticket.addIndividualTicketDetails?.approveStatus === false;

    const notApprovedSystemFlaggingChanges =
      ticket.category?.toString() === GRIEVANCE_CATEGORIES.SYSTEM_FLAGGING &&
      ticket.systemFlaggingTicketDetails?.approveStatus === false;

    const noDuplicatesFound =
      ticket.category?.toString() === GRIEVANCE_CATEGORIES.DEDUPLICATION &&
      !ticket.needsAdjudicationTicketDetails?.selectedIndividual &&
      !ticket.needsAdjudicationTicketDetails?.isMultipleDuplicatesVersion;

    const noDuplicatesFoundMultiple =
      ticket.category?.toString() === GRIEVANCE_CATEGORIES.DEDUPLICATION &&
      ticket.needsAdjudicationTicketDetails?.isMultipleDuplicatesVersion &&
      !ticket.needsAdjudicationTicketDetails?.selectedIndividuals.length;

    // added msg handling for
    let confirmationMessage = '';
    if (notApprovedDeleteIndividualChanges) {
      confirmationMessage = t(
        'You did not approve any changes. Are you sure you want to close the ticket?',
      );
    } else if (notApprovedAddIndividualChanges) {
      confirmationMessage = t('You did not approve any changes.');
    } else if (notApprovedSystemFlaggingChanges) {
      confirmationMessage = '';
    } else if (noDuplicatesFound) {
      confirmationMessage = t(
        'By continuing you acknowledge that individuals in this ticket were reviewed and all were deemed unique to the system. No duplicates were found',
      );
    } else if (noDuplicatesFoundMultiple) {
      confirmationMessage = t(
        'By continuing you acknowledge that individuals in this ticket were reviewed and all were deemed unique to the system. No duplicates were found',
      );
    }
    return confirmationMessage;
  };

  const getClosingConfirmationExtraText = (): string => {
    switch (ticket.issueType?.toString()) {
      case GRIEVANCE_ISSUE_TYPES.EDIT_HOUSEHOLD:
        return getClosingConfirmationExtraTextForIndividualAndHouseholdDataChange();
      case GRIEVANCE_ISSUE_TYPES.EDIT_INDIVIDUAL:
        return getClosingConfirmationExtraTextForIndividualAndHouseholdDataChange();

      default:
        return getClosingConfirmationExtraTextForOtherTypes();
    }
  };

  const closingConfirmationText = t(
    'Are you sure you want to close the ticket?',
  );

  const closingWarningText =
    ticket?.businessArea.postponeDeduplication === true
      ? t(
          'This ticket will be closed without running the deduplication process.',
        )
      : null;

  const changeState = async (status): Promise<void> => {
    try {
      await mutate({
        variables: {
          grievanceTicketId: ticket.id,
          status,
        },
        refetchQueries: () => [
          {
            query: GrievanceTicketDocument,
            variables: { id: ticket.id },
          },
        ],
      });
    } catch (e) {
      e.graphQLErrors.map((x) => showMessage(x.message));
    }
  };

  const getClosingConfirmationText = (): string => {
    if (ticket.category.toString() === GRIEVANCE_CATEGORIES.DEDUPLICATION) {
      return getClosingConfirmationExtraText();
    }
    let additionalContent = '';
    const notApprovedSystemFlaggingChanges =
      ticket.category?.toString() === GRIEVANCE_CATEGORIES.SYSTEM_FLAGGING &&
      ticket.systemFlaggingTicketDetails?.approveStatus === false;

    if (notApprovedSystemFlaggingChanges) {
      additionalContent = t(
        ' By continuing you acknowledge that individuals in this ticket was compared with sanction list. No matches were found',
      );
    }

    const householdHasOneIndividual =
      ticket.household?.activeIndividualsCount === 1;
    if (householdHasOneIndividual) {
      additionalContent = t(
        ' When you close this ticket, the household that this Individual is a member of will be deactivated.',
      );
    }
    return `${closingConfirmationText}${additionalContent}`;
  };

  let closeButton = (
    <Button
      color='primary'
      variant='contained'
      onClick={() =>
        confirm({
          title: t('Close ticket'),
          extraContent:
            ticket.category.toString() === GRIEVANCE_CATEGORIES.DEDUPLICATION
              ? closingConfirmationText
              : getClosingConfirmationExtraText(),
          content: getClosingConfirmationText(),
          warningContent: closingWarningText,
          continueText: t('close ticket'),
        }).then(() => {
          changeState(GRIEVANCE_TICKET_STATES.CLOSED);
        })
      }
      data-cy='button-close-ticket'
    >
      {t('Close Ticket')}
    </Button>
  );
  if (
    ticket.category.toString() === GRIEVANCE_CATEGORIES.DEDUPLICATION &&
    ticket?.needsAdjudicationTicketDetails?.hasDuplicatedDocument &&
    !ticket?.needsAdjudicationTicketDetails?.isMultipleDuplicatesVersion &&
    !!ticket?.needsAdjudicationTicketDetails?.selectedIndividual
  ) {
    closeButton = (
      <ButtonDialog
        title={t('Duplicate Document Conflict')}
        buttonText={t('Close Ticket')}
        message={t(
          'The individuals have matching document numbers. HOPE requires that document numbers are unique. Please resolve before closing the ticket.',
        )}
      />
    );
  }

  if (
    ticket.category.toString() === GRIEVANCE_CATEGORIES.DEDUPLICATION &&
    ticket?.needsAdjudicationTicketDetails?.hasDuplicatedDocument &&
    ticket?.needsAdjudicationTicketDetails?.isMultipleDuplicatesVersion &&
    !!ticket?.needsAdjudicationTicketDetails?.selectedIndividuals.length
  ) {
    closeButton = (
      <ButtonDialog
        title={t('Duplicate Document Conflict')}
        buttonText={t('Close Ticket')}
        message={t(
          'The individuals have matching document numbers. HOPE requires that document numbers are unique. Please resolve before closing the ticket.',
        )}
      />
    );
  }

  const canCreateDataChange = (): boolean => {
    return [
      GRIEVANCE_ISSUE_TYPES.PAYMENT_COMPLAINT,
      GRIEVANCE_ISSUE_TYPES.FSP_COMPLAINT,
    ].includes(ticket.issueType?.toString());
  };

  return (
    <PageHeader
      title={`Ticket ID: ${ticket.unicefId}`}
      breadCrumbs={breadCrumbsItems}
    >
      <Box display='flex' alignItems='center'>
        {isEditable && canEdit && (
          <Box mr={3}>
            <Button
              color='primary'
              variant='outlined'
              component={Link}
              to={`/${businessArea}/grievance-and-feedback/edit-ticket/${id}`}
              startIcon={<EditIcon />}
              data-cy='button-edit'
            >
              {t('Edit')}
            </Button>
          </Box>
        )}
        {isNew && canEdit && canAssign && <Separator />}
        {isNew && canEdit && canAssign && (
          <>
            <LoadingButton
              loading={loading}
              color='primary'
              variant='contained'
              onClick={() => changeState(GRIEVANCE_TICKET_STATES.ASSIGNED)}
            >
              {t('ASSIGN TO ME')}
            </LoadingButton>
          </>
        )}
        {isAssigned && canSetInProgress && (
          <Box mr={3}>
            <LoadingButton
              loading={loading}
              color='primary'
              variant='contained'
              onClick={() => {
                changeState(GRIEVANCE_TICKET_STATES.IN_PROGRESS);
              }}
              data-cy='button-set-to-in-progress'
            >
              {t('Set to in progress')}
            </LoadingButton>
          </Box>
        )}
        {isInProgress && (
          <>
            {canSetOnHold && (
              <Box mr={3}>
                <LoadingButton
                  loading={loading}
                  color='primary'
                  variant='outlined'
                  onClick={() => changeState(GRIEVANCE_TICKET_STATES.ON_HOLD)}
                >
                  {t('Set On Hold')}
                </LoadingButton>
              </Box>
            )}
            {canSendForApproval && (
              <Box mr={3}>
                <LoadingButton
                  loading={loading}
                  color='primary'
                  variant='contained'
                  onClick={() =>
                    changeState(GRIEVANCE_TICKET_STATES.FOR_APPROVAL)
                  }
                  data-cy='button-send-for-approval'
                >
                  {t('Send For Approval')}
                </LoadingButton>
              </Box>
            )}
            {isFeedbackType && canClose && (
              <Button
                color='primary'
                variant='contained'
                onClick={() =>
                  confirm({
                    content: closingConfirmationText,
                    continueText: 'close ticket',
                  }).then(() => changeState(GRIEVANCE_TICKET_STATES.CLOSED))
                }
                data-cy='button-close-ticket'
              >
                {t('Close Ticket')}
              </Button>
            )}
          </>
        )}
        {isOnHold && (
          <>
            {canSetInProgress && (
              <Box mr={3}>
                <LoadingButton
                  loading={loading}
                  color='primary'
                  variant='contained'
                  onClick={() =>
                    changeState(GRIEVANCE_TICKET_STATES.IN_PROGRESS)
                  }
                >
                  {t('Set to in progress')}
                </LoadingButton>
              </Box>
            )}
            {canSendForApproval && (
              <Box mr={3}>
                <LoadingButton
                  loading={loading}
                  color='primary'
                  variant='contained'
                  onClick={() =>
                    changeState(GRIEVANCE_TICKET_STATES.FOR_APPROVAL)
                  }
                  data-cy='button-send-for-approval'
                >
                  {t('Send For Approval')}
                </LoadingButton>
              </Box>
            )}
            {isFeedbackType && canClose && (
              <Button
                color='primary'
                variant='contained'
                onClick={() =>
                  confirm({
                    content: closingConfirmationText,
                    continueText: 'close ticket',
                  }).then(() => changeState(GRIEVANCE_TICKET_STATES.CLOSED))
                }
                data-cy='button-close-ticket'
              >
                {t('Close Ticket')}
              </Button>
            )}
          </>
        )}
        {isForApproval && (
          <>
            {canSendBack && (
              <Box mr={3}>
                <LoadingButton
                  loading={loading}
                  color='primary'
                  variant='contained'
                  onClick={() =>
                    changeState(GRIEVANCE_TICKET_STATES.IN_PROGRESS)
                  }
                >
                  {t('Send Back')}
                </LoadingButton>
              </Box>
            )}
            {canCreateDataChange() && (
              <Box mr={3}>
                <Button
                  onClick={() =>
                    history.push({
                      pathname: `/${businessArea}/grievance-and-feedback/new-ticket`,
                      state: {
                        category: GRIEVANCE_CATEGORIES.DATA_CHANGE,
                        selectedIndividual: ticket.individual,
                        selectedHousehold: ticket.household,
                        linkedTicketId: ticket.id,
                      },
                    })
                  }
                  variant='outlined'
                  color='primary'
                >
                  {t('Create a Data Change ticket')}
                </Button>
              </Box>
            )}
            {canClose && closeButton}
          </>
        )}
      </Box>
    </PageHeader>
  );
};
