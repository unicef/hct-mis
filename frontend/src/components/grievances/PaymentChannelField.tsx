import { Grid, IconButton } from '@material-ui/core';
import { Delete } from '@material-ui/icons';
import { Field } from 'formik';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { FormikTextField } from '../../shared/Formik/FormikTextField';
import { AllIndividualsQuery } from '../../__generated__/graphql';
import { LabelizedField } from '../core/LabelizedField';

export interface PaymentChannelProps {
  index: number;
  baseName: string;
  onDelete: () => {};
  isEdited?: boolean;
  paymentChannel?: AllIndividualsQuery['allIndividuals']['edges'][number]['node']['paymentChannels'][number];
}

export function PaymentChannelField({
  index,
  baseName,
  onDelete,
  isEdited,
  paymentChannel,
}: PaymentChannelProps): React.ReactElement {
  const { t } = useTranslation();
  return (
    <>
      <Grid item xs={11} />
      {!isEdited ? (
        <Grid item xs={1}>
          <IconButton onClick={onDelete}>
            <Delete />
          </IconButton>
        </Grid>
      ) : null}
      <Grid item xs={4}>
        <LabelizedField label={t('Payment channel item')} value='IBAN' />
      </Grid>
      <Grid item xs={4}>
        <LabelizedField
          label={t('Current Value')}
          value={paymentChannel?.bankAccountNumber}
        />
      </Grid>
      <Grid item xs={3}>
        <Field
          name={`${baseName}[${index}].bankAccountNumber`}
          fullWidth
          variant='outlined'
          label={t('New Value')}
          component={FormikTextField}
          required
        />
      </Grid>
      <Grid item xs={4}>
        <LabelizedField
          label={t('Payment channel item')}
          value={t('Bank name')}
        />
      </Grid>
      <Grid item xs={4}>
        <LabelizedField
          label={t('Current Value')}
          value={paymentChannel?.bankName}
        />
      </Grid>
      <Grid item xs={3}>
        <Field
          name={`${baseName}[${index}].bankName`}
          fullWidth
          variant='outlined'
          label={t('New Value')}
          component={FormikTextField}
          required
        />
      </Grid>
    </>
  );
}
