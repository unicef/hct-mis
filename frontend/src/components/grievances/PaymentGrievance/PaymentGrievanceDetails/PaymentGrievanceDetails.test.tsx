import React from 'react';
import { render } from '../../../../testUtils/testUtils';
import { PaymentGrievanceDetails } from '.';
import { fakeApolloApprovePaymentDetailsMutation } from '../../../../../fixtures/grievances/fakeApolloApprovePaymentDetailsMutation';
import { fakeGrievanceTicketPaymentVerification } from '../../../../../fixtures/grievances/fakeGrievanceTicketPaymentVerification';
import { MockedProvider } from '@apollo/react-testing';

describe('components/grievances/PaymentGrievanceDetails', () => {
  it('should render with data', async () => {
    const { container } = render(
      <MockedProvider
        addTypename={false}
        mocks={fakeApolloApprovePaymentDetailsMutation}
      >
        <PaymentGrievanceDetails
          ticket={fakeGrievanceTicketPaymentVerification}
          canApprovePaymentVerification
        />
      </MockedProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
