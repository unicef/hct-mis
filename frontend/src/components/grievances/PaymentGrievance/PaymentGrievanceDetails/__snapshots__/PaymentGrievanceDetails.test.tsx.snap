// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/grievances/PaymentGrievanceDetails should render with data 1`] = `
<div>
  <div
    class="MuiPaper-root sc-gzVnrw fCXiuQ MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="sc-htpNat hUqbZq"
    >
      <div
        class="MuiBox-root MuiBox-root-1"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6"
        >
          Payment Details
        </h6>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Approve
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <table
      class="MuiTable-root sc-htoDjs fmAHNm"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          />
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          >
            Entitlement Value
             ($)
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          >
            Delivered Value
             ($)
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          >
            Received Value
             ($)
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          >
            New Verified Value
             ($)
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body"
          />
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
          />
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
          />
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
          >
            0
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
          >
            45
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
